pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go

```diff
@@ -0,0 +1,223 @@
+}
+
+func (n *ProviderCatalogLauncher) createUpdateExternalDNSApp(obj *v3.GlobalDNSProvider, answers map[string]string) (runtime.Object, error) {
+	//check if provider already running for this GlobalDNSProvider.
+	existingApp, err := n.getProviderIfAlreadyRunning(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	if existingApp != nil {
+		//check if answers should be updated
+		if !reflect.DeepEqual(existingApp.Spec.Answers, answers) {
+			appToupdate := existingApp.DeepCopy()
+			appToupdate.Spec.Answers = answers
+			_, err = n.Apps.Update(appToupdate)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return nil, err
+			}
+		}
+	} else {
+		//create new app
+		appCatalogID := settings.SystemExternalDNSCatalogID.Get()
+		sysProject, err := n.getSystemProjectID()
+		if err != nil {
+			return nil, err
+		}
+
+		//set ownerRef to the dnsProvider CR
+		controller := true
+		ownerRef := []metav1.OwnerReference{{
+			Name:       obj.Name,
+			APIVersion: "v3",
+			UID:        obj.UID,
+			Kind:       obj.Kind,
+			Controller: &controller,
+		}}
+
+		toCreate := pv3.App{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations:     CopyCreatorID(nil, obj.Annotations),
+				Name:            fmt.Sprintf("%s-%s", "systemapp", obj.Name),
+				Namespace:       sysProject,
+				OwnerReferences: ownerRef,
+			},
+			Spec: pv3.AppSpec{
+				ProjectName:     localClusterName + ":" + sysProject,
+				TargetNamespace: namespace.GlobalNamespace,
+				ExternalID:      appCatalogID,
+				Answers:         answers,
+			},
+		}
+		// Now create the App instance
+		_, err = n.Apps.Create(&toCreate)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+func (n *ProviderCatalogLauncher) getProviderIfAlreadyRunning(obj *v3.GlobalDNSProvider) (*pv3.App, error) {
+	sysProject, err := n.getSystemProjectID()
+	if err != nil {
+		return nil, err
+	}
+	existingApp, err := n.appLister.Get(sysProject, fmt.Sprintf("%s-%s", "systemapp", obj.Name))
+
+	if (err != nil && k8serrors.IsNotFound(err)) || existingApp == nil {
+		return nil, nil
+	} else if err != nil && !k8serrors.IsNotFound(err) {
+		logrus.Errorf("GlobaldnsProviderCatalogLauncher: Error listing external-dns %v app %v", obj.Name, err)
+		return nil, err
+	}
+
+	return existingApp, nil
+}
+
+func (n *ProviderCatalogLauncher) getSystemProjectID() (string, error) {
+	systemProject, err := project.GetSystemProject(localClusterName, n.ProjectLister)
+	if err != nil {
+		return "", err
+	}
+
+	return systemProject.Name, nil
+}
+
+func CopyCreatorID(toAnnotations, fromAnnotations map[string]string) map[string]string {
+	if val, exist := fromAnnotations[cattleCreatorIDAnnotationKey]; exist {
+		if toAnnotations == nil {
+			toAnnotations = make(map[string]string, 2)
+		}
+
+		toAnnotations[cattleCreatorIDAnnotationKey] = val
+	}
+	return toAnnotations
+}
+
+func ifAnswersDiffer(existingAnswers map[string]string, newAnswers map[string]string) bool {
```

Comment:
We should remove the useless function.

Reply:
I think we might need to use this instead of reflect.DeepEqual

Reply:
@mrajashree why do you think so? is it better because of performance?

Reply:
@prachidamle because reflect.DeepEqual will require arrays sorted, I forgot to mention this in my above comment but left another comment here https://github.com/rancher/rancher/pull/18240#discussion_r259561462

Reply:
nvm, I thought this is multiclusterapp.Answers which is an array, and didn't realize it's app.Answers